### ISSUE TITLE
esbuild is not throwing error for missing imports/undeclared variables

### ISSUE THREAD

**Issue by Spielboerg** [evanw/esbuild#3470]
I don't know if it's a bug or just technically impossible to check this case, but if variables are not declared/imported, esbuild does not print an error. Example:

https://esbuild.github.io/try/#YgAwLjE5LjUAewogIGJ1bmRsZTogdHJ1ZSwKICBmb3JtYXQ6ICdlc20nLAogIG1pbmlmeTogdHJ1ZQp9AGUAZW50cnkuanMALy8gaW1wb3J0IHsgdGVzdCB9IGZyb20gJy4vZmlsZS5qcycKCmNvbnNvbGUubG9nKHRlc3QuZm9vKQAAZmlsZS5qcwBleHBvcnQgY29uc3QgdGVzdCA9IHsgZm9vOiAnYmFyJyB9

I would expect that esbuild is throwing an error like:
```
✘ [ERROR] "test" is not defined

    entry.js:3:20:
      3 │ console.log(test.foo)
        │             ~~~~
```

**Comment by hyrious** [comment]
In fact, using a name at the top level implicitly means calling `globalThis.{name}`. If there is `globalThis.test` then your example code will not crash. So referencing an unknown name is allowed from the bundler's view.

On the other hand, it is unclear which names are <q>known</q>. For example in the browser environment you may or may not have `CompressionStream`, `VideoDecoder`, etc. It is unable to maintain this list.

You can use toolings like TypeScript or a linter to tell you the missing variables.

**Comment by Spielboerg** [comment]
Thanks for the explanation and the refresher on basic JS knowledge!

We have a linter that, when run, will throw an error in this case. But if we work in serve & watch mode, the linter is not running at this time, so a missing import could be overlooked.

Maybe a `noGlobalThis` or `globalThisAllowList` and `globalThisDisallowList` build options could solve this? Alternatively `implicitAllowList` etc.

**Examples:**
Code:
```js
console.log(foo)
console.log(bar)
```
---
Config 1:
```js
{
  globalThisAllowList: ['foo']
}
```
Would throw an error in the second line: `console.log(bar)`

---

Config 2:
```js
{
  globalThisDisallowList: ['foo']
}
```
Would throw an error in the first line: `console.log(foo)`

---

Config 3:
```js
{
  globalThisAllowList: []  // default is undefined, so if it's an empty array → no implicit names are allowed
}
```
or
```js
{
  noGlobalThis: true
}
```
Would throw an error in both lines.

---

But as you said, the list of names that would need to be added to the `globalThisAllowList` would perhaps be far too long and impossible to maintain. But if the developers could maintain this list themselves and adapt it to their project, it might be useful.

**Comment by evanw** [comment]
> We have a linter that, when run, will throw an error in this case. But if we work in serve & watch mode, the linter is not running at this time, so a missing import could be overlooked.

The solution to not running your linter should not be to turn esbuild into a linter. The solution should be to run your linter. You probably want it to run in parallel to esbuild in watch or serve mode. That can be done by opening a second terminal window and running the linter in watch mode there. Or you could write an esbuild plugin to run it during the build if you need the build to fail on linter errors, but that’s not recommended because it would negatively impact build time. 

I’m closing this issue because turning esbuild into a linter is out of scope. FWIW most people using esbuild are using TypeScript which already handles things like this.
